src/main/utility/status_bar.rs

```diff
@@ -118,6 +114,79 @@ impl<T: 'static + StatusBarState> StatusBar<T> {
     }
 }
 
+pub struct StatusPrinter<T: StatusBarState> {
+    state: Arc<RwLock<T>>,
+    stop_cond: Arc<(Mutex<bool>, Condvar)>,
+    thread: Option<std::thread::JoinHandle<()>>,
+}
+
+impl<T: 'static + StatusBarState> StatusPrinter<T> {
+    /// Create and start drawing the status bar.
+    pub fn new(state: T) -> Self {
+        let state = Arc::new(RwLock::new(state));
+        let stop_cond = Arc::new((Mutex::new(false), Condvar::new()));
+
+        Self {
+            state: Arc::clone(&state),
+            stop_cond: Arc::clone(&stop_cond),
```

Comment:
Optional nit - a "one shot channel" (`std::sync::mpsc::channel<()>`) might be a little easier to maintain than these lower level sync primitives. The print loop can use `recv_timeout` with a 30s timeout. On timeout print the status. When something is received, exit.

Reply:
Good idea, I've replaced the mutex/condvar with a channel.